### ISSUE TITLE
out of bounds error when first coordinate of point is negative

### ISSUE THREAD

**Issue by hjy1210** [NSoiffer/MathCAT#199]
There are two lines in [this web page](https://hjy1210.github.io/nvda/mathcat/mathcat-outofbound-pi.html), both lines have issue.

About the issue(NVDA+MathCat) of first line,

* speech viewer window shows:

```en
In        the interval from  0  to  2   π          including  0  but not including  2   π        
, how many solutions of equation
        f         of          x         is equal to  1 half        ? 
```

* user hear voice as:

```en
In        the interval from 0 to 2 including 0 but not including 2 how many solutions of 
equation f of x is equal to  1 half 
```

**Notice that π is missing in voice.**

About the second line, there are three points, we(NVDA+MathCat) **CANNOT** browse enter point C:

When firefox load the page, NVDA log window shows error:

```en
ERROR - queueHandler.flushQueue (10:41:15.268) - MainThread (20828):
Error in func _TextReader.nextLine
Traceback (most recent call last):
  File "queueHandler.pyc", line 64, in flushQueue
  File "speech\sayAll.pyc", line 297, in nextLine
  File "speech\types.pyc", line 50, in _flattenNestedSequences
  File "speech\types.pyc", line 50, in <genexpr>
  File "speech\speech.pyc", line 1538, in getTextInfoSpeech
  File "speech\speech.pyc", line 1203, in _extendSpeechSequence_addMathForTextInfo
  File "C:\Users\hjy\AppData\Roaming\nvda\addons\MathCAT\globalPlugins\MathCAT\MathCAT.py", line 367, in getSpeechForMathMl
    return ConvertSSMLTextForNVDA(libmathcat.GetSpokenText(), self._language)
pyo3_runtime.PanicException: index out of bounds: the len is 1 but the index is 1
```

When enter the first line, NVDA log window shows error:
```en
ERROR - scriptHandler.executeScript (10:43:03.045) - MainThread (20828):
error executing script: <bound method CursorManager.script_moveByLine_forward of <virtualBuffers.gecko_ia2.Gecko_ia2 object at 0x01229D50>> with gesture 'down arrow'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 295, in executeScript
  File "cursorManager.pyc", line 274, in script_moveByLine_forward
  File "cursorManager.pyc", line 171, in _caretMovementScriptHelper
  File "speech\speech.pyc", line 1231, in speakTextInfo
  File "speech\types.pyc", line 42, in __iter__
  File "speech\speech.pyc", line 1538, in getTextInfoSpeech
  File "speech\speech.pyc", line 1203, in _extendSpeechSequence_addMathForTextInfo
  File "C:\Users\hjy\AppData\Roaming\nvda\addons\MathCAT\globalPlugins\MathCAT\MathCAT.py", line 367, in getSpeechForMathMl
    return ConvertSSMLTextForNVDA(libmathcat.GetSpokenText(), self._language)
pyo3_runtime.PanicException: index out of bounds: the len is 1 but the index is 1
```

When trying enter the point C, NVDA log window shows error:

```en
ERROR - scriptHandler.executeScript (10:45:12.433) - MainThread (20828):
error executing script: <bound method CursorManager.script_moveByCharacter_forward of <virtualBuffers.gecko_ia2.Gecko_ia2 object at 0x01229D50>> with gesture 'right arrow'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 295, in executeScript
  File "cursorManager.pyc", line 261, in script_moveByCharacter_forward
  File "cursorManager.pyc", line 171, in _caretMovementScriptHelper
  File "speech\speech.pyc", line 1231, in speakTextInfo
  File "speech\types.pyc", line 42, in __iter__
  File "speech\speech.pyc", line 1402, in getTextInfoSpeech
  File "speech\speech.pyc", line 1203, in _extendSpeechSequence_addMathForTextInfo
  File "C:\Users\hjy\AppData\Roaming\nvda\addons\MathCAT\globalPlugins\MathCAT\MathCAT.py", line 367, in getSpeechForMathMl
    return ConvertSSMLTextForNVDA(libmathcat.GetSpokenText(), self._language)
pyo3_runtime.PanicException: index out of bounds: the len is 1 but the index is 1
```

**Comment by andrew-l-d** [comment]
When I read that page NVDA (within my limited knowledge) reads correctly, including the "p".  However, when trying to activate MathCAT I receive the "Not math" error.
I checked the page with the [New Validation Service ](http://validator.w3.org/nu/
) and errors were reported.  I therefore copied the MathML into a blank page and everything seemed to read correctly, including both MathML elements.

Andrew

**Comment by hjy1210** [comment]
@andrew-l-d 
Thank for point me to the New Html checker, the resport of that web page is:
1. Following three meta tags should have no trailing slash
```
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
```
2. and need a title tag.

After correct in my local file, upload to the checker for checking, report no error, the isseus I mentioned are still there.

**Comment by andrew-l-d** [comment]
hjy1210, which synthesizer are you using?  When using eSpeak or Eloquence, the second pi is presented for me.  When using Windows Onecore Voice it is not.  Math content is being recognised though.
I am still finding your original page.  I suspect I am missing some information as pi is only mentioned once.  I am also not finding information such as 3 marks.  I am getting more information in Chrome, but suspect still not reading everything.

Andrew

**Comment by hjy1210** [comment]
@andrew-l-d, 
About synthesizer, I use Windows OneCore voices, even change to Microsoft Speech API version 5, the issues remains.
I can not understand those voices of eSpeak NG.
I have no Eloquence option.

**Comment by NSoiffer** [comment]
Thanks for the example page. That makes it easy to debug. Fix is coming soon.

**Comment by NSoiffer** [comment]
[A beta build of MathCAT with this fix (and others) can be found here](https://github.com/NSoiffer/MathCATForPython/releases/download/v.0.3.7-beta.4/MathCAT-0.3.7.nvda-addon).

Please let me know if you find any problems.

**Comment by hjy1210** [comment]
@NSoiffer 
First of all, `MathCAT-0.37.nvda-addon` DO solve the issue about `out of bounds error` which sometimes can make NVDA go crazy.

The other issue which I mentioned in the first line of my example page is still there although it is not an issue about the title of issue #199. 

For that issue I made three tests, which hope can help understanding the problem. Difference between tests is the attribute width value of mspace tag. Only the third test is correct, which mean we can hear the sound of `pi`. I wonder why the mspace width values(0.277em/0.278em) cause problem.

First test:
```
#[test]
fn parens_interval_closed_open2() {
    let expr = "<math>
	<mrow>
		<mo stretchy=\"false\" form=\"prefix\">[</mo>
		<mspace width=\"0.333em\"></mspace>
		<mn>0</mn>
		<mspace width=\"0.333em\"></mspace>
		<mo>,</mo>
		<mspace width=\"0.333em\"></mspace>
		<mn>2</mn>
		<mi>π</mi>
		<mspace width=\"0.333em\"></mspace>
		<mo stretchy=\"false\" form=\"postfix\">)</mo>
	</mrow>
</math>";
test("en", "SimpleSpeak",expr, "the closed open interval from 0 to 2 π");
}
```
the result:
```
running 1 test
thread 'Languages::en::SimpleSpeak::functions::parens_interval_closed_open2' panicked at tests\common\mod.rs:43:23:
assertion `left == right` failed:
test with en/SimpleSpeak failed
  left: "the closed open interval from 0 to 2 π"
 right: "the closed open interval from \u{a0}0\u{a0} to \u{a0}2 π"
```

Second test:
```
#[test]
fn parens_interval_closed_open3() {
    let expr = "<math>
	<mrow>
		<mo stretchy=\"false\" form=\"prefix\">[</mo>
		<mspace width=\"0.278em\"></mspace>
		<mn>0</mn>
		<mspace width=\"0.278em\"></mspace>
		<mo>,</mo>
		<mspace width=\"0.278em\"></mspace>
		<mn>2</mn>
		<mi>π</mi>
		<mspace width=\"0.278em\"></mspace>
		<mo stretchy=\"false\" form=\"postfix\">)</mo>
	</mrow>
</math>";
test("en", "SimpleSpeak",expr, "the closed open interval from 0 to 2 π");
}
```
result:
```
running 1 test
thread 'Languages::en::SimpleSpeak::functions::parens_interval_closed_open3' panicked at tests\common\mod.rs:43:23:
assertion `left == right` failed:
test with en/SimpleSpeak failed
  left: "the closed open interval from 0 to 2 π"
 right: "the closed open interval from \u{a0}0\u{a0} to \u{a0}2 π"
```

Third test:
```
#[test]
fn parens_interval_closed_open4() {
    let expr = "<math>
	<mrow>
		<mo stretchy=\"false\" form=\"prefix\">[</mo>
		<mspace width=\"0.277em\"></mspace>
		<mn>0</mn>
		<mspace width=\"0.277em\"></mspace>
		<mo>,</mo>
		<mspace width=\"0.277em\"></mspace>
		<mn>2</mn>
		<mi>π</mi>
		<mspace width=\"0.277em\"></mspace>
		<mo stretchy=\"false\" form=\"postfix\">)</mo>
	</mrow>
</math>";
test("en", "SimpleSpeak",expr, "the closed open interval from 0 to 2 π");
}
```
result:
```
running 1 test
thread 'Languages::en::SimpleSpeak::functions::parens_interval_closed_open4' panicked at tests\common\mod.rs:43:23:
assertion `left == right` failed:
test with en/SimpleSpeak failed
  left: "the closed open interval from 0 to 2 π"
 right: "the closed open interval from 0 to 2 pi"
```

**Comment by NSoiffer** [comment]
Sorry, I overlooked the "pi" issue. I think I know why it might be happening: spaces are moved into `mi` and `mn` for various reasons. Identifiers greater than length 1 are presumed to be words and are spoken, thus π  doesn't become "pi".  I have code to strip the spaces off, but obviously something isn't working.

**Comment by NSoiffer** [comment]
This turned out to be very subtle and exposed a bug in MathCAT's rule system involving when strings run through the Unicode mappings and when they don't. It is now fixed and all the tests pass, but the solution involves a little bit of built-in hacks as exposing this decision to the rule writer.
